core/handler/convert_fields.go

```diff
@@ -176,3 +180,121 @@ func RunUnsafeCode(vm *otto.Otto, code string, timeOut time.Duration) (value ott
+	vm.Set("payload", payload)
+	// Value will have the returning value of the Encoder function (otto.Value)
+	value, err := RunUnsafeCode(vm, fmt.Sprintf("(%s)(payload)", f.Encoder), timeOut)
+	if err != nil {
+		return nil, errors.New("Error while running the JS function")
+	}
+
+	//not necessary --> use IsObject method
+	if value.Class() != "Array" {
+		return nil, errors.New("Encoder should return an Array.")
+	}
+
+	// Export will attempt to convert the value to a Go representation and return it via an interface{} kind
+	v, err := value.Export()
+	if err != nil {
+		return nil, fmt.Errorf("Error while encoding: %s", err)
+	}
+
+	m, ok := v.([]int64)
+	if !ok {
+		return nil, errors.New("Encoder should return an Array of numbers")
+	}
+
+	res := make([]byte, len(m))
+	for i, v := range m {
+		if v < 0 || v > 255 {
+			return nil, errors.New("Encoder error: numbers in array should be between 0 and 255")
+		}
+		res[i] = byte(v)
+	}
+
+	return res, nil
+}
+
+// ProcessDownLink encode the data sent by the application in []bytes which can be sent to the device
+// The payload is interface{} because it can be JSON or []bytes for instance (and maybe other kind of type later)
+func (f *DownlinkFunctions) ProcessDownlink(payload map[string]interface{}) ([]byte, bool, error) {
+	encoded, err := f.Encode(payload)
+	if err != nil {
+		return nil, false, err
+	}
+	// We make the hypothesis that the handler as just to encode data without checking if it's valid or not.
+	return encoded, true, nil
+}
+
+// We have to have a ttnDown *pb_broker.DownlinkMessage argument because of the type func in downlink.go
+func (h *handler) ConvertFieldsDown(ctx log.Interface, appDown *mqtt.DownlinkMessage, ttnDown *pb_broker.DownlinkMessage) error {
+
+	// The validity of the message is the application responsability.
+	// Message is the payload in []byte coming from the conversion of the Fields
+	// Checking the fields of the message coming from the App -> MQTT
+	if appDown.Fields != nil {
+
+		ctx.Debug("HERE 1") // To delete
+
+		// Find Application thanks to the AppID, so we can access the Payload functions of the Applications
+		app, err := h.applications.Get(appDown.AppID)
+		if err != nil {
+			ctx.Debug("HERE 2") // To delete
+			return nil          // Do not process if application not found
+		}
+
+		ctx.Debug("HERE 3") // To delete
+
+		// Downlink processor -> Downlink fields
+		functions := &DownlinkFunctions{
+			Encoder: app.Encoder,
+		}
+
+		ctx.Debug("HERE 4") // To delete
+
+		message, valid, err := functions.ProcessDownlink(appDown.Fields)
+		if err != nil {
+			return errors.New("Error while processing")
+		}
+
+		// Impossible to have fields and payload at the same time
+		if appDown.Payload != nil {
+			return errors.New("Error processing downlink: both Fields and Payload provided")
+		}
+
+		if !valid {
+			return errors.New("ttn/handler: The processed payload is not valid")
```

Comment:
We actually do not check validity for downlink fields, so this is an obsolete check